File: clinical/src/main/java/com/hartwig/actin/clinical/curation/config/OncologicalHistoryConfigFactory.java
package com.hartwig.actin.clinical.curation.config;

import java.util.Map;

import com.hartwig.actin.clinical.curation.CurationUtil;
import com.hartwig.actin.clinical.datamodel.ImmutablePriorTumorTreatment;
import com.hartwig.actin.clinical.datamodel.PriorTumorTreatment;
import com.hartwig.actin.clinical.interpretation.TreatmentCategoryResolver;
import com.hartwig.actin.util.ResourceFile;

import org.jetbrains.annotations.NotNull;

public class OncologicalHistoryConfigFactory implements CurationConfigFactory<OncologicalHistoryConfig> {

    @NotNull
    @Override
    public OncologicalHistoryConfig create(@NotNull Map<String, Integer> fields, @NotNull String[] parts) {
        boolean ignore = CurationUtil.isIgnoreString(parts[fields.get("name")]);
        return ImmutableOncologicalHistoryConfig.builder()
                .input(parts[fields.get("input")])
                .ignore(ignore)
                .curated(!ignore ? curateObject(fields, parts) : null)
                .build();
    }

    @NotNull
    private static PriorTumorTreatment curateObject(@NotNull Map<String, Integer> fields, @NotNull String[] parts) {
        return ImmutablePriorTumorTreatment.builder()
                .name(parts[fields.get("name")])
                .startYear(ResourceFile.optionalInteger(parts[fields.get("startYear")]))
                .startMonth(ResourceFile.optionalInteger(parts[fields.get("startMonth")]))
                .stopYear(ResourceFile.optionalInteger(parts[fields.get("stopYear")]))
                .stopMonth(ResourceFile.optionalInteger(parts[fields.get("stopMonth")]))
                .bestResponse(ResourceFile.optionalString(parts[fields.get("bestResponse")]))
                .stopReason(ResourceFile.optionalString(parts[fields.get("stopReason")]))
                .categories(TreatmentCategoryResolver.fromStringList(parts[fields.get("category")]))
                .isSystemic(ResourceFile.bool(parts[fields.get("isSystemic")]))
                .chemoType(ResourceFile.optionalString(parts[fields.get("chemoType")]))
                .immunoType(ResourceFile.optionalString(parts[fields.get("immunoType")]))
                .targetedType(ResourceFile.optionalString(parts[fields.get("targetedType")]))
                .hormoneType(ResourceFile.optionalString(parts[fields.get("hormoneType")]))
                .radioType(ResourceFile.optionalString(parts[fields.get("radioType")]))
                .carTType(ResourceFile.optionalString(parts[fields.get("carTType")]))
                .transplantType(ResourceFile.optionalString(parts[fields.get("transplantType")]))
                .supportiveType(ResourceFile.optionalString(parts[fields.get("supportiveType")]))
                .trialAcronym(ResourceFile.optionalString(parts[fields.get("trialAcronym")]))
                .build();
    }
}

File: clinical/src/main/java/com/hartwig/actin/clinical/curation/config/OncologicalHistoryConfigFactory.kt
package com.hartwig.actin.clinical.curation.config

import com.hartwig.actin.clinical.datamodel.treatment.ImmutablePriorTumorTreatment

class OncologicalHistoryConfigFactory : CurationConfigFactory<OncologicalHistoryConfig> {
    override fun create(fields: Map<String?, Int?>, parts: Array<String>): OncologicalHistoryConfig {
        val ignore: Boolean = CurationUtil.isIgnoreString(parts[fields["name"]!!])
        return ImmutableOncologicalHistoryConfig.builder()
            .input(parts[fields["input"]!!])
            .ignore(ignore)
            .curated(if (!ignore) curateObject(fields, parts) else null)
            .build()
    }

    companion object {
        private fun curateObject(fields: Map<String?, Int?>, parts: Array<String>): PriorTumorTreatment {
            return ImmutablePriorTumorTreatment.builder()
                .name(parts[fields["name"]!!])
                .startYear(ResourceFile.optionalInteger(parts[fields["startYear"]!!]))
                .startMonth(ResourceFile.optionalInteger(parts[fields["startMonth"]!!]))
                .stopYear(ResourceFile.optionalInteger(parts[fields["stopYear"]!!]))
                .stopMonth(ResourceFile.optionalInteger(parts[fields["stopMonth"]!!]))
                .cycles(ResourceFile.optionalInteger(parts[fields["cycles"]!!]))
                .bestResponse(ResourceFile.optionalString(parts[fields["bestResponse"]!!]))
                .stopReason(ResourceFile.optionalString(parts[fields["stopReason"]!!]))
                .categories(TreatmentCategoryResolver.fromStringList(parts[fields["category"]!!]))
                .isSystemic(ResourceFile.bool(parts[fields["isSystemic"]!!]))
                .chemoType(ResourceFile.optionalString(parts[fields["chemoType"]!!]))
                .immunoType(ResourceFile.optionalString(parts[fields["immunoType"]!!]))
                .targetedType(ResourceFile.optionalString(parts[fields["targetedType"]!!]))
                .hormoneType(ResourceFile.optionalString(parts[fields["hormoneType"]!!]))
                .radioType(ResourceFile.optionalString(parts[fields["radioType"]!!]))
                .carTType(ResourceFile.optionalString(parts[fields["carTType"]!!]))
                .transplantType(ResourceFile.optionalString(parts[fields["transplantType"]!!]))
                .supportiveType(ResourceFile.optionalString(parts[fields["supportiveType"]!!]))
                .trialAcronym(ResourceFile.optionalString(parts[fields["trialAcronym"]!!]))
                .ablationType(ResourceFile.optionalString(parts[fields["ablationType"]!!]))
                .build()
        }
    }
}